File: moi-dotnet/MoiBackend.Infrastructure/Services/ReportService.cs
using Microsoft.EntityFrameworkCore;
using NPOI.XWPF.UserModel;
using MoiBackend.Core.Interfaces;
using MoiBackend.Infrastructure.Data;

namespace MoiBackend.Infrastructure.Services;

public class ReportService : IReportService
{
    private readonly MoiDbContext _dbContext;

    private const string HeaderColor = "2B5797";   // Dark blue
    private const string AltRowColor = "EEF2FF";  // Light blue
    private const string TotalRowColor = "DBEAFE"; // Soft blue for total
    private const string White = "FFFFFF";

    public ReportService(MoiDbContext dbContext)
    {
        _dbContext = dbContext;
    }

    public async Task<byte[]> GenerateOverallWordReportAsync(long eventId)
    {
        var moiTransactions = await _dbContext.MoiTransactions
            .AsNoTracking()
            .Where(t => t.EventId == eventId)
            .Include(t => t.Contributor)
            .ToListAsync();

        var goldEntries = await _dbContext.GoldEntries
            .AsNoTracking()
            .Where(g => g.EventId == eventId)
            .Include(g => g.Contributor)
            .ToListAsync();

        // Sort both lists by village name (ascending, case-insensitive)
        moiTransactions = moiTransactions
            .OrderBy(tx => tx.Contributor?.Village ?? "", StringComparer.OrdinalIgnoreCase)
            .ToList();

        goldEntries = goldEntries
            .OrderBy(ge => ge.Contributor?.Village ?? "", StringComparer.OrdinalIgnoreCase)
            .ToList();

        using var doc = new XWPFDocument();

        // =====================================================
        // MOI COLLECTION REPORT
        // =====================================================
        AddSectionTitle(doc, "Moi Collection Report");

        if (moiTransactions.Count > 0)
        {
            string[] moiHeaders = { "S.No", "Name", "Village", "Amount (Rs)" };
            var moiTable = doc.CreateTable(1, moiHeaders.Length);
            moiTable.Width = 5000;
            StyleHeaderRow(moiTable.GetRow(0), moiHeaders);

            decimal grandTotal = 0;
            int sno = 1;
            foreach (var tx in moiTransactions)
            {
                grandTotal += tx.Amount;
                var row = moiTable.CreateRow();
                bool isAlt = (sno % 2 == 0);
                SetCellText(row.GetCell(0), (sno++).ToString(), isAlt, true);
                SetCellText(row.GetCell(1), tx.Contributor?.Name ?? "", isAlt, false);
                SetCellText(row.GetCell(2), tx.Contributor?.Village ?? "", isAlt, false);
                SetCellText(row.GetCell(3), "Rs. " + tx.Amount.ToString("0.00"), isAlt, true);
            }

            // Grand Total row
            var totalRow = moiTable.CreateRow();
            SetCellWithColor(totalRow.GetCell(0), "", TotalRowColor, false);
            SetCellWithColor(totalRow.GetCell(1), "", TotalRowColor, false);
            SetCellWithColor(totalRow.GetCell(2), "GRAND TOTAL :", TotalRowColor, true);
            SetCellWithColor(totalRow.GetCell(3), "Rs. " + grandTotal.ToString("0.00"), TotalRowColor, true);
        }
        else
        {
            AddNoteText(doc, "No Moi entries found for this event.");
        }

        // Page break
        var breakPara = doc.CreateParagraph();
        breakPara.CreateRun().AddBreak(BreakType.PAGE);

        // =====================================================
        // GOLD COLLECTION REPORT
        // =====================================================
        AddSectionTitle(doc, "Gold Collection Report");

        if (goldEntries.Count > 0)
        {
            string[] goldHeaders = { "S.No", "Name", "Village", "Gold Details" };
            var goldTable = doc.CreateTable(1, goldHeaders.Length);
            goldTable.Width = 5000;
            StyleHeaderRow(goldTable.GetRow(0), goldHeaders);

            int sno2 = 1;
            foreach (var ge in goldEntries)
            {
                var row = goldTable.CreateRow();
                bool isAlt = (sno2 % 2 == 0);
                SetCellText(row.GetCell(0), (sno2++).ToString(), isAlt, true);
                SetCellText(row.GetCell(1), ge.Contributor?.Name ?? "", isAlt, false);
                SetCellText(row.GetCell(2), ge.Contributor?.Village ?? "", isAlt, false);
                SetCellText(row.GetCell(3), ge.GoldDetails, isAlt, false);
            }
        }
        else
        {
            AddNoteText(doc, "No Gold entries found for this event.");
        }

        using var ms = new MemoryStream();
        doc.Write(ms);
        return ms.ToArray();
    }

    private static void AddSectionTitle(XWPFDocument doc, string titleText)
    {
        var para = doc.CreateParagraph();
        para.Alignment = ParagraphAlignment.CENTER;
        para.SpacingBefore = 200;
        para.SpacingAfter = 200;
        var run = para.CreateRun();
        run.SetText(titleText);
        run.IsBold = true;
        run.FontSize = 16;
        run.SetColor(HeaderColor);
        run.FontFamily = "Calibri";
    }

    private static void AddNoteText(XWPFDocument doc, string note)
    {
        var para = doc.CreateParagraph();
        var run = para.CreateRun();
        run.SetText(note);
        run.IsItalic = true;
        run.SetColor("888888");
        run.FontFamily = "Calibri";
    }

    private static void StyleHeaderRow(XWPFTableRow headerRow, string[] headers)
    {
        for (int i = 0; i < headers.Length; i++)
        {
            var cell = headerRow.GetCell(i);
            cell.SetColor(HeaderColor);
            var para = cell.Paragraphs[0];
            para.Alignment = ParagraphAlignment.CENTER;
            ClearRuns(para);
            var run = para.CreateRun();
            run.SetText(headers[i]);
            run.IsBold = true;
            run.SetColor(White);
            run.FontSize = 11;
            run.FontFamily = "Calibri";
        }
    }

    private static void SetCellText(XWPFTableCell cell, string text, bool isAlt, bool centered)
    {
        if (isAlt)
        {
            cell.SetColor(AltRowColor);
        }
        var para = cell.Paragraphs[0];
        para.Alignment = centered ? ParagraphAlignment.CENTER : ParagraphAlignment.LEFT;
        ClearRuns(para);
        var run = para.CreateRun();
        run.SetText(text ?? "");
        run.FontSize = 10;
        run.FontFamily = "Calibri";
    }

    private static void SetCellWithColor(XWPFTableCell cell, string text, string color, bool centered)
    {
        cell.SetColor(color);
        var para = cell.Paragraphs[0];
        para.Alignment = centered ? ParagraphAlignment.CENTER : ParagraphAlignment.LEFT;
        ClearRuns(para);
        var run = para.CreateRun();
        run.SetText(text ?? "");
        run.IsBold = true;
        run.FontSize = 11;
        run.FontFamily = "Calibri";
    }

    private static void ClearRuns(XWPFParagraph para)
    {
        for (int r = para.Runs.Count - 1; r >= 0; r--)
        {
            para.RemoveRun(r);
        }
    }
}

File: moi/src/main/java/com/moi/service/ReportService.java
package com.moi.service;

import com.moi.model.GoldEntry;
import com.moi.model.MoiTransaction;
import com.moi.repository.GoldEntryRepository;
import com.moi.repository.MoiTransactionRepository;
import lombok.RequiredArgsConstructor;
import org.apache.poi.xwpf.usermodel.*;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.math.BigDecimal;
import java.util.Comparator;
import java.util.List;

@Service
@RequiredArgsConstructor
public class ReportService {

    private final MoiTransactionRepository moiTransactionRepository;
    private final GoldEntryRepository goldEntryRepository;

    // Colors (hex strings for high-level POI API)
    private static final String HEADER_COLOR = "2B5797";   // Dark blue
    private static final String ALT_ROW_COLOR = "EEF2FF";  // Light blue
    private static final String TOTAL_ROW_COLOR = "DBEAFE"; // Soft blue for total
    private static final String WHITE = "FFFFFF";

    @Transactional(readOnly = true)
    public byte[] generateOverallWordReport(Long eventId) throws IOException {
        List<MoiTransaction> moiTransactions = moiTransactionRepository.findByEventId(eventId);
        List<GoldEntry> goldEntries = goldEntryRepository.findByEventId(eventId);

        // Sort both lists by village name (ascending, case-insensitive)
        moiTransactions.sort(Comparator.comparing(
                tx -> tx.getContributor().getVillage(), String.CASE_INSENSITIVE_ORDER));
        goldEntries.sort(Comparator.comparing(
                ge -> ge.getContributor().getVillage(), String.CASE_INSENSITIVE_ORDER));

        try (XWPFDocument document = new XWPFDocument()) {

            // =====================================================
            // MOI COLLECTION REPORT
            // =====================================================
            addSectionTitle(document, "Moi Collection Report");

            if (!moiTransactions.isEmpty()) {
                String[] moiHeaders = {"S.No", "Name", "Village", "Amount (Rs)"};
                XWPFTable moiTable = document.createTable(1, moiHeaders.length);
                moiTable.setWidth("100%");
                styleHeaderRow(moiTable.getRow(0), moiHeaders);

                BigDecimal grandTotal = BigDecimal.ZERO;
                int sno = 1;
                for (MoiTransaction tx : moiTransactions) {
                    grandTotal = grandTotal.add(tx.getAmount());
                    XWPFTableRow row = moiTable.createRow();
                    boolean isAlt = (sno % 2 == 0);
                    setCellText(row.getCell(0), String.valueOf(sno++), isAlt, true);
                    setCellText(row.getCell(1), tx.getContributor().getName(), isAlt, false);
                    setCellText(row.getCell(2), tx.getContributor().getVillage(), isAlt, false);
                    setCellText(row.getCell(3), "Rs. " + grandTotal.toString(), isAlt, true);
                }

                // Grand Total row
                XWPFTableRow totalRow = moiTable.createRow();
                setCellWithColor(totalRow.getCell(0), "", TOTAL_ROW_COLOR, false);
                setCellWithColor(totalRow.getCell(1), "", TOTAL_ROW_COLOR, false);
                setCellWithColor(totalRow.getCell(2), "GRAND TOTAL :", TOTAL_ROW_COLOR, true);
                setCellWithColor(totalRow.getCell(3), "Rs. " + grandTotal.toString(), TOTAL_ROW_COLOR, true);

            } else {
                addNoteText(document, "No Moi entries found for this event.");
            }

            // Page break
            document.createParagraph().createRun().addBreak(BreakType.PAGE);

            // =====================================================
            // GOLD COLLECTION REPORT
            // =====================================================
            addSectionTitle(document, "Gold Collection Report");

            if (!goldEntries.isEmpty()) {
                String[] goldHeaders = {"S.No", "Name", "Village", "Gold Details"};
                XWPFTable goldTable = document.createTable(1, goldHeaders.length);
                goldTable.setWidth("100%");
                styleHeaderRow(goldTable.getRow(0), goldHeaders);

                int sno2 = 1;
                for (GoldEntry ge : goldEntries) {
                    XWPFTableRow row = goldTable.createRow();
                    boolean isAlt = (sno2 % 2 == 0);
                    setCellText(row.getCell(0), String.valueOf(sno2++), isAlt, true);
                    setCellText(row.getCell(1), ge.getContributor().getName(), isAlt, false);
                    setCellText(row.getCell(2), ge.getContributor().getVillage(), isAlt, false);
                    setCellText(row.getCell(3), ge.getGoldDetails(), isAlt, false);
                }
            } else {
                addNoteText(document, "No Gold entries found for this event.");
            }

            ByteArrayOutputStream out = new ByteArrayOutputStream();
            document.write(out);
            return out.toByteArray();
        }
    }

    // ─── Helper: Big section title ──────────────────────────────────────────
    private void addSectionTitle(XWPFDocument document, String titleText) {
        XWPFParagraph para = document.createParagraph();
        para.setAlignment(ParagraphAlignment.CENTER);
        para.setSpacingBefore(200);
        para.setSpacingAfter(200);
        XWPFRun run = para.createRun();
        run.setText(titleText);
        run.setBold(true);
        run.setFontSize(16);
        run.setColor(HEADER_COLOR);
        run.setFontFamily("Calibri");
    }

    // ─── Helper: Italic grey note ───────────────────────────────────────────
    private void addNoteText(XWPFDocument document, String note) {
        XWPFParagraph para = document.createParagraph();
        XWPFRun run = para.createRun();
        run.setText(note);
        run.setItalic(true);
        run.setColor("888888");
        run.setFontFamily("Calibri");
    }

    // ─── Helper: Style header row (blue background + white bold text) ────────
    private void styleHeaderRow(XWPFTableRow headerRow, String[] headers) {
        for (int i = 0; i < headers.length; i++) {
            XWPFTableCell cell = headerRow.getCell(i);
            cell.setColor(HEADER_COLOR);
            XWPFParagraph para = cell.getParagraphs().get(0);
            para.setAlignment(ParagraphAlignment.CENTER);
            clearRuns(para);
            XWPFRun run = para.createRun();
            run.setText(headers[i]);
            run.setBold(true);
            run.setColor(WHITE);
            run.setFontSize(11);
            run.setFontFamily("Calibri");
        }
    }

    // ─── Helper: Cell with alternating background ───────────────────────────
    private void setCellText(XWPFTableCell cell, String text, boolean isAlt, boolean centered) {
        if (isAlt) {
            cell.setColor(ALT_ROW_COLOR);
        }
        XWPFParagraph para = cell.getParagraphs().get(0);
        para.setAlignment(centered ? ParagraphAlignment.CENTER : ParagraphAlignment.LEFT);
        clearRuns(para);
        XWPFRun run = para.createRun();
        run.setText(text == null ? "" : text);
        run.setFontSize(10);
        run.setFontFamily("Calibri");
    }

    // ─── Helper: Cell with a specific background color and bold text ─────────
    private void setCellWithColor(XWPFTableCell cell, String text, String color, boolean centered) {
        cell.setColor(color);
        XWPFParagraph para = cell.getParagraphs().get(0);
        para.setAlignment(centered ? ParagraphAlignment.CENTER : ParagraphAlignment.LEFT);
        clearRuns(para);
        XWPFRun run = para.createRun();
        run.setText(text == null ? "" : text);
        run.setBold(true);
        run.setFontSize(11);
        run.setFontFamily("Calibri");
    }

    // ─── Helper: Remove all runs from a paragraph ────────────────────────────
    private void clearRuns(XWPFParagraph para) {
        for (int r = para.getRuns().size() - 1; r >= 0; r--) {
            para.removeRun(r);
        }
    }
}
